src/Data/Graph.purs

```diff
@@ -72,6 +78,29 @@ toMap (Graph g) = g
 vertices :: forall k v. Graph k v -> List v
 vertices (Graph g) = map fst (M.values g)
 
+type EdgesState k v =
+  { unvisited :: List (Tuple k (Tuple v (List k)))
+  , result :: List (Edge k)
+  }
+
+-- | List all edges in a graph
+edges :: forall k v. Graph k v -> List (Edge k)
+edges (Graph g) = go initialState
+  where
+    go :: EdgesState k v -> List (Edge k)
+    go { unvisited: Nil, result: res } = res
+    go { unvisited: (src /\ (_ /\ dests)) : ns, result: res } =
+      go
+        { unvisited: ns
+        , result: map (Edge <<< (src /\ _)) dests <> res
+        }
+
+    initialState :: EdgesState k v
+    initialState =
+      { unvisited: M.toUnfoldableUnordered g
+      , result: Nil
+      }
+
```

Comment:
I'm pretty sure this is slow for a few reasons. 

First, you're using `List` and you have at least two traversals over `dests`:
1. `map (Edge <<< (src /\ _)) dests`: 1 traversal to create a new version of the list
2. `(map _ dests) <> res`: 1 traversal appending the result of `map _ dests` to `res`. See https://github.com/JordanMartinez/purescript-jordans-reference/blob/latestRelease/21-Hello-World/05-Application-Structure/src/03-Free/01-What-Is-The-Free-Monad/04-The-Remorseless-Free-Monad.md#the-direction-matters for context.

A faster solution would to fold over `dests` and cons the `src` and `destination` onto `res` in one fold:
```purs
result: foldl (\acc dest -> Edge (src /\ dest) : acc) res dests
```

Second, for `initialState`, why convert the `Map` to a `List` in the first place via [`M.toUnfoldableUnordered`](https://github.com/purescript/purescript-ordered-collections/blob/v3.0.0/src/Data/Map/Internal.purs#L642-L650)? That traverses the map once and then you traverse the map a second time in `go` via the `unvisited` field. Since `Graph` is just a newtype around `Map` and `Map` has an instance for `FoldableWithIndex`, couldn't this same key-value information be obtained with `foldlWithIndex` using a single traversal rather than two as it currently does?

Reply:
Wow! I did not notice that at all. Thank you for taking the time to explain this all to me so clearly :)

As you mentioned I've replaced the custom function with `foldlWithIndex` (which traverses the `Map` once) and a `foldl` for each list of edge endpoints. So I cannot see any redundant traversals now.